### DISCUSSION TITLE
Temperature dependent mobility and grain growth simulation

### DISCUSSION THREAD

**Discussion by ykvishal** [idaholab/moose#17433] [Q&A Modules: Phase field] [answered]
**Brief details:**
I am simulating the effect of temperature gradient on grain growth. I modified [3D_6000_gr.i](https://github.com/idaholab/moose/blob/next/modules/phase_field/examples/grain_growth/3D_6000_gr.i) by following [temperature_gradient.i](https://github.com/idaholab/moose/blob/next/modules/phase_field/test/tests/grain_growth/temperature_gradient.i) file. A typical set up is given here:
![Picture11](https://user-images.githubusercontent.com/31565781/112631041-aefb1200-8e36-11eb-9eef-f688e3400172.png)

2D simulation video is shown here:
https://user-images.githubusercontent.com/31565781/112631151-cf2ad100-8e36-11eb-94a7-4e3c6746a1ab.mp4

**Problem:**

I need 5 nodes (X28 cores, pmem=8gb) to run [3D_6000_gr.i](https://github.com/idaholab/moose/blob/next/modules/phase_field/examples/grain_growth/3D_6000_gr.i) (ie 140 cores). But, temperature dependent mobility case needs **560** cores. Also dt is almost 1/10 th of normal grain growth case.  First I started with 40 nodes (X14 cores, pmem=17gb) then I followed memory requirement per core using [VectorMemoryUsage](https://mooseframework.inl.gov/source/vectorpostprocessors/VectorMemoryUsage.html) and I could reduce number of nodes to 28 nodes (X20 cores, pmem=12gb). **However, I still need 560 cores.** . I tried with 280 cores - 20 nodes (X14 cores, pmem=17gb but simulation does not start. Here is the comparison of  [VectorMemoryUsage](https://mooseframework.inl.gov/source/vectorpostprocessors/VectorMemoryUsage.html) 
Physical Memory in GBs
![PhysicalMemory](https://user-images.githubusercontent.com/31565781/112632306-3bf29b00-8e38-11eb-8056-08e725bcf847.jpg)

Node Utilization:
![NodeUtilization](https://user-images.githubusercontent.com/31565781/112632342-4ca31100-8e38-11eb-96d8-e91807b75e5d.jpg)
Here **t** represents Time Step.

I also have [PerfGraphOutput ](https://mooseframework.inl.gov/source/outputs/PerfGraphOutput.html)data.

**For  [3D_6000_gr.i](https://github.com/idaholab/moose/blob/next/modules/phase_field/examples/grain_growth/3D_6000_gr.i)** 
Performance Graph:
![image](https://user-images.githubusercontent.com/31565781/112632852-eff42600-8e38-11eb-8264-3cd08ca2d576.png)
Heaviest Branch:
![image](https://user-images.githubusercontent.com/31565781/112632967-131ed580-8e39-11eb-8bff-78db1cd2549e.png)
Heaviest Sections:
![image](https://user-images.githubusercontent.com/31565781/112633054-2a5dc300-8e39-11eb-9a2d-c92af530a2c6.png)

-------------------------------------------------------------------------------------------------------------------------------
**For Temperature dependent mobility case:**
Performance Graph:
![image](https://user-images.githubusercontent.com/31565781/112633344-888aa600-8e39-11eb-8c2d-e100090aa42b.png)

Heaviest Branch:
![image](https://user-images.githubusercontent.com/31565781/112633404-9c360c80-8e39-11eb-8dad-8495b37c8625.png)

Heaviest Sections:
![image](https://user-images.githubusercontent.com/31565781/112633469-ad7f1900-8e39-11eb-9d9f-3a23a500a78b.png)

Is there a way to reduce number of cores to run a temperature dependent mobility case? Also, is there a way to increase **dt**?
Thank you.

Edit: I made minor changes to the  [3D_6000_gr.i](https://github.com/idaholab/moose/blob/next/modules/phase_field/examples/grain_growth/3D_6000_gr.i) file like 1) added parallel_type = distributed to the mesh block; 2) Neumann boundary condition ; and 3) changed op_num to 50 from 28 due to problem with GrainTracker. I will create a separate thread on this GrainTracker issue.

**Comment by laagesen** [comment, marked as answer]
If you are changing the temperature profile, the rate of grain coarsening is going to change. If the temperature is lower and coarsening slows down, if you are using mesh adaptivity (and you should for a problem this large) you are going to have more degrees of freedom in your problem. So, it would be helpful to know how many degrees of freedom are in the problem for the two cases. Can you add a `NumDOFs` postprocessor to your input files and run? If there are dramatically more DOF's in the problem with the gradient there may not be much you can do other than add more cores. However just want to find out if this is the case.

**Reply by ykvishal** [reply]
Thank you @laagesen for your reply. 
For 3D_6000_gr.i,  the post-processor output is as follows

![image](https://user-images.githubusercontent.com/31565781/112894710-7bd9ac80-90dc-11eb-9d47-8dc53ce3cef8.png)


For Temperature dependent mobility case:,  the post-processor output is as follows
![image](https://user-images.githubusercontent.com/31565781/112894603-59479380-90dc-11eb-805f-b2465433f3ec.png)

**Reply by laagesen** [reply]
OK the number of DOF's is not changing in either case, so you are not using adaptivity. (I'm not sure why you have a smaller number of DOF's in the temperature-dependent problem- from the n_elements postprocessor it looks like the mesh size is different right? I would stick to the same mesh size if possible so you can make an apples-to-apples comparison). Anyways, I think you should use mesh adaptivity for a large problem like this. This should reduce the number of DOF's and therefore the number of cores you need. You can see a good example of how this is done in the example `moose/modules/phase_field/examples/grain_growth/grain_growth_2D_voronoi_newadapt.i`
and you can find documentation on the adaptivity system here:
https://mooseframework.inl.gov/syntax/Adaptivity/index.html

**Reply by laagesen** [reply]
Also, please make sure all parameters are set the same between the two input files other than the temperature gradient. I just noticed that the `wGB` parameter in the `GBEvolution` material in the two input files that you started from are very different from each other, if those are different values in your different cases, the performance will be very different. `wGB` controls the width of the diffuse interface that represents grain boundaries.

**Reply by ykvishal** [reply]
Yes, I kept everything else same from [3D_6000_gr.i](https://github.com/idaholab/moose/blob/next/modules/phase_field/examples/grain_growth/3D_6000_gr.i) and I only copied temperature related part from [temperature_gradient.i](https://github.com/idaholab/moose/blob/next/modules/phase_field/test/tests/grain_growth/temperature_gradient.i)
My materials block is as follows
[Materials]
  [./Copper]
    type = GBEvolution
    T = T #T = 500
    wGB = 3 # um
    GBmob0 = 2.5e-6 #m^4/(Js) from Schoenfelder 1997
    Q = 0.23 #Migration energy in eV
    GBenergy = 0.708 #GB energy in J/m^2
    molar_volume = 7.11e-6 #Molar volume in m^3/mol
    length_scale = 1.0e-6
    time_scale = 1.0
  [../]
[]

I already tried adaptivity in 2D. I will try it with 3D simulations and see what happens.

**Comment by mnabilbhuiyan** [comment]
Hello @laagesen, in this [temperature_gradient.i](https://github.com/idaholab/moose/blob/next/modules/phase_field/test/tests/grain_growth/temperature_gradient.i) input file in the
 [Functions]
  [./TGradient]
    type = ParsedFunction
    expression = '450 + 0.1*x'
  [../]
[]
section instead of x, if I use t, the expression will be then  '450 + 0.1*x'. In this case, will it indicate the temperature change is time-dependent and also rate of grain growth will be time dependent?_

**Reply by laagesen** [reply]
yes you should be able to substitute t for x in that ParsedFunction. Just be aware grain growth could accelerate a lot at later times and convergence of the problem will get slower.

**Reply by mnabilbhuiyan** [reply]
Thank you so much for clearing this.
